# MetaVar - Exploring genomic variation in metagenomic datasets

## Workflow
* Pipeline Script
  * ###### Inputs
    * subparser for the entry phase of the files into the pipeline
	  * 0 = start = run everything
        * gene file(s) [.bed or .gff]
        * multi-sample .vcf file
        * output directory
        * file with the sample pairings
	  * 1 = skip get_paired_vcf
	    * gene files(s) [.bed or .gff]
		* output directory
		* .txt files to be used
	  * 2 = skip extract_genomic_diff
	    * gene file(s) [.bed or .gff]
		* output directory
		* .diff files to be used
	  * 3 = skip aggregate_genomic-diff
	    * TBD
  * ###### Execution
    * Runs the above specified
	  * 0
	    * get_paired_vcf
		* extract_genomic_diff
		* aggregate_genomic_diff
	  * 1
	    * extract_genomic_diff
		*aggregate_genomic_diff
	  * 2
	    * aggregate_genomic_diff
	  * 3
	    * TBD
	* main script : run this script with the inputs for the other scripts
	  * scheduler script will run each operation for each file
	    * this "file" could be a pairwise, or it could be a pairwise for each genome used too
	* Hierarchy:
		* main (config script)
		  * scheduler - for each comparison
		  	* get_paired_vcf
			* extract_genomic_diff
			* aggregate_genomic_diff
			* merge_genomic_coord
	* Example:
		* run main and specify 5 pairwise samples
		* specify 8 genomes to compare
		  * output 16 files (the .tbl and .comb files)
		* for each genome
		  * for each pair
		    * merge
			* extract
	      * for each extracted
		    * aggregate
		* could make this work by adding an additional parameter to the files which says read a specific config file
		  that is generated by the main script and is updated with the next file to be used after each script is run
		* could optionally use argparse (use an import into the main script to identify a boolean use_parser) and when
		  not using could instead just import the next scripts and have run in that manner
  * ###### Structure
      * need to look at with Ranjit
	  * Parallelization
	    * look like a for loop of the main command?
  * ###### Output
    * the outputs of the individual files above
	* able to set the deletion of intermediate files

* Need to add a parameter to the script that includes parsing the config file for the input instead

##### Will have to move some files around at this point
* python get_paired_vcf.py example/Bacteroides_vulgatus.vcf.snp.gatk D11 R11
* python get_paired_vcf.py example/Bacteroides_vulgatus.vcf.snp.gatk D8 R8
* python get_paired_vcf.py example/Bacteroides_vulgatus.vcf.snp.gatk D16 R16

* python extract_genomic_diff.py D11_R11.txt
* python extract_genomic_diff.py D8_R8.txt
* python extract_genomic_diff.py D16_R16.txt

* python merge_genes.py example/D8_R8.txt.diff example/D11_R11.txt.diff example/D16_R16.txt.diff 

* python aggregate_genomic_diff.py example/D8_R8_D11_R11_D16_R16.txt.tbl example/Bacteroides_vulgatus.gff


## Workflow Description
1. extract pairwise vcf files
* extract the SNP differences
* merge individual .diff files into a table
* using a gff or bed file, lookup in what genes the SNP differences are occuring and output to a combined file

* output a file for GeneBrowser viewing
* do analysis

